filebeat/input/filestream/internal/input-logfile/harvester_test.go

```diff
@@ -0,0 +1,331 @@
+			// harvester is waiting to start
+			_, ok := hg.readers.table[source.Name()]
+			if ok && locked {
+				releaseResource(r)
+				locked = false
+			}
+		})
+
+		// wait until harvester.Run is done
+		wg.Wait()
+		// wait until goroutine that started `harvester.Run` is finished
+		gorountineChecker.waitOriginalCount()
+		require.Equal(t, 1, mockHarvester.runCount)
+		require.Nil(t, hg.StopGroup())
+	})
+
+	t.Run("assert already locked resource has no problem when harvestergroup is cancelled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: correctOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+		gorountineChecker := newGoroutineChecker()
+
+		r, err := lock(inputCtx, hg.store, source.Name())
+		if err != nil {
+			t.Fatalf("cannot lock source")
+		}
+		defer releaseResource(r)
+
+		hg.Start(inputCtx, source)
+
+		gorountineChecker.waitWhileMoreGoroutinesWithFunc(func() {
+			err := hg.StopGroup()
+			require.Error(t, err)
+		})
+
+		require.Equal(t, 0, mockHarvester.runCount)
+
+		gorountineChecker.waitOriginalCount()
+	})
+}
+
+func testDefaultHarvesterGroup(t *testing.T, mockHarvester Harvester) *defaultHarvesterGroup {
+	return &defaultHarvesterGroup{
+		readers:   newReaderGroup(),
+		pipeline:  &pipelinemock.MockPipelineConnector{},
+		harvester: mockHarvester,
+		store:     testOpenStore(t, "test", nil),
+		tg:        unison.TaskGroup{},
+	}
+}
+
+type gorountineChecker struct {
+	n    int
+	wait time.Duration
+}
+
+func newGoroutineChecker() *gorountineChecker {
+	return &gorountineChecker{
+		n:    runtime.NumGoroutine(),
+		wait: 10 * time.Millisecond,
+	}
+}
+
+func (c *gorountineChecker) waitWhileMoreGoroutinesWithFunc(f func()) {
+	for c.n < runtime.NumGoroutine() {
+		fmt.Println("waiting with function", c.n, runtime.NumGoroutine())
+		time.Sleep(c.wait)
+
+		f()
+	}
+}
+
+func (c *gorountineChecker) waitOriginalCount() {
+	fmt.Println("waiting until original", c.n, runtime.NumGoroutine())
+	for c.n < runtime.NumGoroutine() {
+		time.Sleep(c.wait)
+	}
+}
+
+type mockHarvester struct {
+	runCount int
+	wg       *sync.WaitGroup
+	onRun    func(input.Context, Source, Cursor, Publisher) error
+}
+
+func (m *mockHarvester) Run(ctx input.Context, s Source, c Cursor, p Publisher) error {
+	if m.wg != nil {
+		defer m.wg.Done()
+	}
+
+	m.runCount += 1
+	if m.onRun != nil {
+		return m.onRun(ctx, s, c, p)
+	}
+	return nil
+}
+
+func (m *mockHarvester) Test(_ Source, _ input.TestContext) error { return nil }
+
+func (m *mockHarvester) Name() string { return "mock" }
+
+func correctOnRun(_ input.Context, _ Source, _ Cursor, _ Publisher) error {
+	return nil
+}
+
+func blockUntilCancelOnRun(c input.Context, _ Source, _ Cursor, _ Publisher) error {
+	<-c.Cancelation.Done()
+	return nil
+}
+
+func errorOnRun(_ input.Context, _ Source, _ Cursor, _ Publisher) error {
+	return fmt.Errorf("harvester error")
+}
```

Comment:
alternatively:

```
type harvesterRun func(input.Context, Source, Cursor, Publisher) error

func harvesterReturnsWith(err error) harvesterRun {
  func(input.Context, Source, Cursor, Publisher) error {
    return err
  }
}
```

now you can you `harvesterReturnsWith` in case of `errorOnRun` or `correctOnRun`. As the error return is defined with the test you can actually test that the error is returned to the owner of the harvester group.